My Notebooks Page - SEAB-5114

KimberleyChong opened the pull request:
**Description**
Adds the `my-notebooks` page (SEAB-5114), adds the GitHub apps registration interface for notebooks (SEAB-5119), and updates and hides notebooks in collections (SEAB-5386).

~~Opening this PR early just to get some feedback earlier rather than later, but when the release branch is made I'll switch the destination branch from `develop` to `release`.~~

**Quick Overview:**


https://user-images.githubusercontent.com/97123241/232588061-53ecb4a0-927f-4260-84f0-396448847d9b.mp4

What my-notebooks looks like, I just copied the existing notebook in the db and assigned it to myself for the purpose of having a notebook entry:

![image](https://user-images.githubusercontent.com/97123241/232594390-cbf47bc7-7bbe-41e3-8167-95627cf471b0.png)

I created the `register-github-app-modal.component` which closely resembles the register a workflow github apps option:

![image](https://user-images.githubusercontent.com/97123241/232594756-1764ca10-564e-4619-8fb0-c4a44c321943.png)

I also reused the same registration screen for services since previously clicking "Add a service" would open GitHub with no intermediate step.

![image](https://user-images.githubusercontent.com/97123241/232596118-6a4f13b2-27f5-4fe4-956f-3dc59b45e481.png)

- Renamed `SharedWorkflowServicesModule` to `SharedWorkflowServicesNotebooksModule` since it includes all three now
- I removed `MyServicesService` and `MyBioWorkflowsService` components, as they both contained identical functions differing in entry type, and altered `selectEntry` in `MyWorkflowsService` to handle workflows, services and notebooks without calling different `entryType` specific functions.

Also added the warning banner to the collections page along with the feature flag to hide notebooks in collections. The entry link here also needed to be updated to properly navigate to a notebook entry.

![image](https://user-images.githubusercontent.com/97123241/232596422-9f5a6342-897f-4b6d-b829-c05d91e9fe2d.png)

While doing that I decided to re-export the notebook icon, to a smaller size, I noticed the tools and workflows icons have a much smaller "intrinsic size", so I re-exported it to match so we don't need to keep using styles to resize it smaller.


**Review Instructions**
Using the feature flag, my notebooks should appear in the my-sidebar and you should be able to view the my-notebooks page. Everything should appear as the demo video.

**Issue**
SEAB-5114, SEAB-5119, SEAB-5386

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

denis-yuen commented:
> Opening this PR early just to get some feedback earlier rather than later, but when the release branch is made I'll switch the destination branch from `develop` to `release`.

Think this should stay as `develop` actually, if we're targeting it at 1.15

coverbeck commented on src/app/workflow/register-workflow-modal/register-github-app-modal/register-github-app-modal.component.ts:
```diff
@@ -0,0 +1,58 @@
+/*
+ *    Copyright 2018 OICR
```
Out of date copied copyright :)

coverbeck commented on src/app/organizations/collection/collection.component.html:
```diff
@@ -101,18 +104,28 @@ <h3 *ngIf="Dockstore.FEATURES.enableNotebooks">Workflows, Notebooks, and Tools</
+                    '/' +
+                      (entry.entryType === entryType.Tool
+                        ? 'container'
+                        : entry.entryType === entryType.Workflow
```
IDE error on `entryType.Workflow` -- it doesn't exist.

KimberleyChong replied:
Oh woah thanks for catching this, supposed to be `BioWorkflow`.  What IDE do you use Charles? That's really useful

coverbeck replied:
I use WebStorm, but if you're using IntelliJ, I think you can configure it with the same plugins as WebStorm.

coverbeck commented on src/app/organizations/collection/collection.component.html:
```diff
@@ -101,18 +104,28 @@ <h3 *ngIf="Dockstore.FEATURES.enableNotebooks">Workflows, Notebooks, and Tools</
+                    '/' +
+                      (entry.entryType === entryType.Tool
+                        ? 'container'
+                        : entry.entryType === entryType.Workflow
```
IDE error on `entryType.Workflow` -- it doesn't exist. Actually, we have a pipe for this src/app/entry/router-link.pipe.ts

KimberleyChong replied:
> Actually, we have a pipe for this src/app/entry/router-link.pipe.ts

Mm previously considered that, unfortunately it takes in a type of `Workflow` to compute the url but here we're dealing with type `CollectionEntry`. But maybe its a good idea to add support for both, I'll update the pipe to take in `entry: Workflow | CollectionEntry` and handle both cases

coverbeck reviewed (commented):
> Think this should stay as develop actually, if we're targeting it at 1.15

I'm unclear, since we haven't cut the release branches, will this part of 1.14 then (behind the feature flag)? I think it's good if we do, but can go either way.

svonworl commented on src/app/workflow/register-workflow-modal/register-github-app-modal/register-github-app-modal.component.ts:
```diff
@@ -0,0 +1,58 @@
+
+  ngOnInit() {
+    this.username$ = this.userQuery.username$;
+    this.usernameChangeRequired$ = this.userQuery.user$.pipe(map((user) => user.usernameChangeRequired));
```
Definitely not a biggie, but in the interest of consistency, we could make `isNameChangeRequired` an observable property of `userQuery`, and remove the dup code from this and the other modal.

KimberleyChong replied:
Makes sense

svonworl reviewed (approved):
Good job, great work!

denis-yuen commented:
> I'm unclear, since we haven't cut the release branches, will this part of 1.14 then (behind the feature flag)? I think it's good if we do, but can go either way.

Think we just have to keep this open until later in the week when we start cutting release branches

KimberleyChong commented on src/app/myworkflows/myworkflows.service.ts:
```diff
@@ -91,10 +91,31 @@ export class MyWorkflowsService extends MyEntriesService<Workflow, OrgWorkflowOb
-          this.selectEntry(this.recomputeWhatEntryToSelect([...(services || []), ...[]]), EntryType.Service);
         },
         (error) => {
           this.workflowService.setWorkflows([]);
```
Yes this is true 👍️, I can call it with an empty array for the error case

kathy-t commented on src/app/organizations/collection/collection.component.html:
```diff
@@ -101,18 +104,28 @@ <h3 *ngIf="Dockstore.FEATURES.enableNotebooks">Workflows, Notebooks, and Tools</
-              <mat-card class="pb-3">
+              <mat-card
+                *ngIf="
+                  (entry.entryType !== entryType.Notebook && !Dockstore.FEATURES.enableNotebooks) || Dockstore.FEATURES.enableNotebooks
```
Why do you need this `*ngIf`? If it's not a notebook, then it'll satisfy the condition both when `enableNotebooks` is true and false. If it is a notebook, then `enableNotebooks` should be true, which also satisfies the condition

KimberleyChong replied:
> If it is a notebook, then `enableNotebooks` should be true

Yeah the thing is, this entry being a `notebook` and `enableNotebooks` being set are independent of each other, the issue was without this `*ngIf` entries that were notebooks we being shown regardless of the value of `enableNotebooks` (i.e. all the time)

The conditions here are a bit hard to imagine:

The first part:
`entry.entryType !== entryType.Notebook && !Dockstore.FEATURES.enableNotebooks`

- When the feature flag is **not** enabled, display an entry only if it is **not** a notebook

The second part:
`Dockstore.FEATURES.enableNotebooks`

- When the feature flag **is** enabled, display **all** entries (including notebooks)

That being said, the logic here can be simplified and maintain the same results :) will update to:
`entry.entryType !== entryType.Notebook || Dockstore.FEATURES.enableNotebooks`

![image](https://user-images.githubusercontent.com/97123241/233179190-df8e8591-00b7-4d49-a650-2385b18172a6.png)

kathy-t commented on src/app/organizations/collection/collection.component.html:
```diff
@@ -31,6 +31,9 @@
 <app-loading [loading]="(loadingCollection$ | async) || (loadingOrganization$ | async)">
   <div *ngIf="organization$ | async as organization">
     <div class="container mb-5" *ngIf="collection$ | async as collection" fxLayout="column">
+      <div *ngIf="Dockstore.FEATURES.enableNotebooks" class="mt-3">
```
Looking at the collection page in your video, it's not clear what the banner is referring. Perhaps we should mention the feature in the banner, like Charles suggested during the demo.

KimberleyChong replied:
Noted, saw your comment on the original Jira ticket and while doing those updates will add in a `feature` parameter

kathy-t commented on src/app/organizations/collection/collection.component.html:
```diff
@@ -101,18 +104,28 @@ <h3 *ngIf="Dockstore.FEATURES.enableNotebooks">Workflows, Notebooks, and Tools</
 
             <div data-cy="collectionEntry" *ngFor="let entry of collection?.entries">
-              <mat-card class="pb-3">
+              <mat-card
```
Brainstorming, does the notebook mat card need the `preview` flag?

KimberleyChong replied:
Maybe instead of adding it to the card, I could add the preview bubble to the title since we usually add it on the tab/title, similar to my-starred?

<img src="https://user-images.githubusercontent.com/97123241/233199217-bff85d7e-c39c-4aa8-86f3-5962dd27f0bd.png" width=400px>

![image](https://user-images.githubusercontent.com/97123241/233197587-9185e2a0-0893-43ae-ae35-fc8685914e14.png)

kathy-t replied:
I think it's a little ambiguous beside the title because it also mentions workflows and tools. Since you're adding the feature name to the feature warning banner, we can probably live without the preview label on this page

KimberleyChong replied:
Got it, makes sense

kathy-t commented on src/app/workflow/register-workflow-modal/register-github-app-modal/register-github-app-modal.component.html:
```diff
@@ -0,0 +1,33 @@
+    [href]="gitHubAppInstallationLink$ | async"
+    [disabled]="usernameChangeRequired$ | async"
+    matTooltip="Manage Dockstore installations on GitHub"
+    ><mat-icon>add</mat-icon> Manage Dockstore installations on GitHub</a
```
I think this link might have a leading space (I remember seeing something similar on QA)

kathy-t replied:
Yeah, that's the right one. When you hover over the link, you can see an underline for the space, which IMO looks awkward:
![image](https://user-images.githubusercontent.com/25287123/233200360-d7e22c5e-ff3c-4007-87ed-fa3884023ede.png)


Does it make a huge difference without the space? If not, I would remove it. If it does, consider adding a right margin to the icon instead of the space. 

Admittedly, this is a very minor thing so your call whether or not you think it's worth doing :slightly_smiling_face:

KimberleyChong replied:
Ah I see, in that case I can add the `no-underline` style which gets rid of the underline on hover which I think looks better for links styled as buttons :)

kathy-t commented on src/app/myworkflows/my-workflow/my-workflow.component.html:
```diff
@@ -152,7 +138,7 @@ <h3>Shared with me</h3>
                 <ul>
-                  <li *ngIf="entryType === EntryType.Service">
+                  <li *ngIf="entryType === EntryType.Service || entryType === EntryType.Notebook">
                     Press the 'Manage on GitHub' button and follow the GitHub apps web interface to create services from your GitHub
```
This might be out of date because I don't see a Manage on GitHub button

KimberleyChong replied:
I remember thinking it didn't make sense, thanks for verifying. Updating to be as so, which I think makes sense:

```
<li>
  Press the 'Register a {{ entryType$ | async | titlecase }}' button 
  <span *ngIf="entryType === EntryType.Service || entryType === EntryType.Notebook; else manualRegistration">and follow the GitHub apps web interface to create {{ entryType$ | async }}s from your GitHub repositories.</span>
  <ng-template #manualRegistration >to manually add a {{ entryType$ | async }} using the registration wizard.</ng-template>
</li>
```
**For Services/Notebooks:**

Before:
![image](https://user-images.githubusercontent.com/97123241/233190911-81a62839-24eb-43fd-b337-b213e1e1e72b.png)

After:
![image](https://user-images.githubusercontent.com/97123241/233190996-4a82276c-931a-48fa-8d95-414acbb6a7e8.png)

**For workflows:**

Before:
![image](https://user-images.githubusercontent.com/97123241/233192987-4fe597de-a212-420a-924a-7f12fb7d12dd.png)

After:
![image](https://user-images.githubusercontent.com/97123241/233192420-ea9bcf61-1155-485b-bff3-4bfafc29d597.png)

Going to update `my-tool.component` to match while here:
Before:
![image](https://user-images.githubusercontent.com/97123241/233193657-ae71e19c-b89f-46d3-98bd-95de9f871c5a.png)
After:
![image](https://user-images.githubusercontent.com/97123241/233194959-80570c2d-f447-4d45-baed-2dc4a83e1c20.png)

Changed buttons to say "Register a {{ entryType }}" to match the other registration buttons:
![image](https://user-images.githubusercontent.com/97123241/233195089-66caff9a-4c82-4c2b-8465-242caa6ceaba.png)
![image](https://user-images.githubusercontent.com/97123241/233195128-ad80f966-e826-4464-a26b-b545a5dbb95b.png)

kathy-t commented on src/app/workflow/register-workflow-modal/register-github-app-modal/register-github-app-modal.component.html:
```diff
@@ -0,0 +1,33 @@
+<div class="mb-3" *ngIf="Dockstore.FEATURES.enableNotebooks && entryType === EntryType.Notebook">
```
This dialog should have a title, like Register Workflow

KimberleyChong replied:
Thank you!!

KimberleyChong commented on src/app/myworkflows/myworkflows.service.ts:
```diff
@@ -91,10 +91,31 @@ export class MyWorkflowsService extends MyEntriesService<Workflow, OrgWorkflowOb
-          this.selectEntry(this.recomputeWhatEntryToSelect([...(services || []), ...[]]), EntryType.Service);
         },
         (error) => {
           this.workflowService.setWorkflows([]);
```
It's there :) just cut off in the diff:

https://github.com/dockstore/dockstore-ui2/blob/12d6858b460398eb511f48cf0827fcbc3f26b51c/src/app/myworkflows/myworkflows.service.ts#L123

KimberleyChong commented on src/app/myworkflows/sidebar-accordion/sidebar-accordion.component.html:
```diff
@@ -63,11 +63,14 @@
+              <p
+                [attr.data-cy]="'no-published-' + pageEntryType + '-message'"
+                *ngIf="
+                  pageEntryType === EntryType.Service || pageEntryType === EntryType.BioWorkflow || pageEntryType === EntryType.Notebook
```
Yeah, `apptools` isn't included because the format `No published {{ pageEntryType }}s.` doesn't fit well. 

The `pageEntryType` for an appTool is `appTool`, which displays `No published appTools.` instead of `No published tools.` so I decided to keep it in a case of it's own

But great point! I can actually simplify a lot and remove the `*ngIf`:
```
<p [attr.data-cy]="'no-published-' + pageEntryType + '-message'" class="text-center size-small">
  No published {{ pageEntryType === EntryType.Service || pageEntryType === EntryType.BioWorkflow || pageEntryType === EntryType.Notebook ? pageEntryType : 'tool' }}s.
</p>
```

svonworl replied:
We also now have the `EntryTypeMetadata` objects, which contain the singular and plural terms that we use to publicly refer to each entry type.  It may not be the right time to do so, but if/when we convert this template to use them, we may not even need the conditional.

KimberleyChong commented on src/app/myworkflows/myworkflows.service.ts:
```diff
@@ -213,11 +246,30 @@ export class MyWorkflowsService extends MyEntriesService<Workflow, OrgWorkflowOb
+   * @param entryType
+   */
+  openRegisterGithubAppModal(entryType: EntryType) {
+    const dialogRef = this.matDialog.open(RegisterGithubAppModalComponent, { width: '600px', data: entryType });
```
Makes sense, I'll go with `bootstrap4mediumModalSize` and update for both `RegisterWorkflowModal` and `RegisterGithubAppModal` then

KimberleyChong commented on src/app/myworkflows/sidebar-accordion/sidebar-accordion.component.html:
```diff
@@ -94,11 +97,14 @@
+                "
+                class="text-center size-small"
+              >
+                No unpublished {{ pageEntryType }}s.
```
Gotcha

KimberleyChong replied:
Should services have it as well then since they're also registered through GitHub apps? I supposed they're excluded for a reason though

denis-yuen replied:
Probably never got around to it, probably not a priority still

KimberleyChong commented:
> > Think this should stay as develop actually, if we're targeting it at 1.15
> 
> I'm unclear, since we haven't cut the release branches, will this part of 1.14 then (behind the feature flag)? I think it's good if we do, but can go either way.

Yep, everything is hidden behind the feature flag, no strong feelings whether it makes the release or not, but I suppose one thing that would be good to have in production from this PR is the feature flag to hide notebooks in collections... otherwise notebooks appear if an organization has added it to their collection without the FF, but also it leads to a 404 if you click on it because the link also gets updated in this PR.

KimberleyChong commented:
**Update**: Changes following feedback, main points:
- Updated `routerLinkPipe` to support `collectionEntry` and `DockstoreTool` types to use for routing collections
- added `isUsernameChangeRequired$` as an observable boolean to `userQuery` for consistency/to remove dup subscription code
- Use Bootstrap for modal size, add header to modal and other updates to the `register {{ entryType }}` modals
- Simplify logic, update outdated registration text

denis-yuen commented:
FYI this is fine to go into develop now from a github perspective

kathy-t commented on src/app/myworkflows/sidebar-accordion/sidebar-accordion.component.html:
```diff
@@ -60,17 +60,13 @@
+              <p [attr.data-cy]="'no-published-' + pageEntryType + '-message'" class="text-center size-small">
+                No published
+                {{
                   pageEntryType === EntryType.Service || pageEntryType === EntryType.BioWorkflow || pageEntryType === EntryType.Notebook
```
This could be simplified to `pageEntryType == EntryType.AppTool ? 'tool' : pageEntryType`

kathy-t commented on src/app/myworkflows/sidebar-accordion/sidebar-accordion.component.html:
```diff
@@ -94,25 +90,23 @@
+              <p [attr.data-cy]="'no-unpublished-' + pageEntryType + '-message'" class="text-center size-small">
+                No unpublished
+                {{
                   pageEntryType === EntryType.Service || pageEntryType === EntryType.BioWorkflow || pageEntryType === EntryType.Notebook
```
Same comment

kathy-t commented on src/app/entry/router-link.pipe.ts:
```diff
@@ -1,25 +1,71 @@
-      case EntryType.Notebook:
-        return '/notebooks/' + workflow.full_workflow_path;
+
+    if ((entry as Workflow).full_workflow_path) {
```
Steve mentioned that we now have `EntryTypeMetadata` in this [comment](https://github.com/dockstore/dockstore-ui2/pull/1756#discussion_r1171580682) and I think using that here would greatly simplify this function. `EntryTypeMetadata.sitePath` contains the entry type specific portion of the site path (ex: `containers`), so that would remove the need for all the switches.

`EntryTypeMetadata` should be a field in `DockstoreTool | Workflow`, so you can easily access it there. As for `CollectionEntry`, it's not a field in there, but you can use this function in the `EntryTypeMetadataService` to retrieve the `EntryTypeMetadata` based on the collection entry's `entryType`: https://github.com/dockstore/dockstore-ui2/blob/3478bfb2e97481e1b127224bdb06b268ef8c94a1/src/app/entry/type-metadata/entry-type-metadata.service.ts#L30

KimberleyChong commented:
**Update**:
- refactor `RouterLinkPipe` to use `EntryTypeMetadata` and update `spec.ts`
- add `EntryTypeMetadata` to mock objects in `openapi-mocked-objects.ts`
- create `EntryTypeMetadataStubService` in `service-stubs.ts`
